renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,607 @@
+VkImageLayout VulkanDebugManager::GetImageLayout(ResourceId image, VkImageAspectFlags aspect,
+                                                 uint32_t mip)
+{
+  VkImageLayout imgLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+  const ImageLayouts &imgLayouts = m_pDriver->m_ImageLayouts[image];
+  for(const auto &resState : imgLayouts.subresourceStates)
+  {
+    VkImageSubresourceRange range = resState.subresourceRange;
+
+    if((range.aspectMask & aspect) &&
+       (mip >= range.baseMipLevel && mip < range.baseMipLevel + range.levelCount))
+    {
+      // Consider using the layer count
+      imgLayout = resState.newLayout;
+    }
+  }
+  return imgLayout;
+}
+
+std::vector<PixelModification> VulkanReplay::PixelHistory(std::vector<EventUsage> events,
+                                                          ResourceId target, uint32_t x, uint32_t y,
+                                                          uint32_t slice, uint32_t mip,
+                                                          uint32_t sampleIdx, CompType typeHint)
+{
+  std::vector<PixelModification> history;
+  VkResult vkr;
+  VkDevice dev = m_pDriver->GetDev();
+
+  if(events.empty())
+    return history;
+
+  VulkanCreationInfo::Image imginfo = GetDebugManager()->GetImageInfo(target);
+  if(imginfo.format == VK_FORMAT_UNDEFINED)
+    return history;
+
+  // TODO: figure out correct aspect.
+  VkImageLayout imgLayout = GetDebugManager()->GetImageLayout(target, VK_IMAGE_ASPECT_COLOR_BIT, mip);
+  RDCASSERTNOTEQUAL(imgLayout, VK_IMAGE_LAYOUT_UNDEFINED);
+
+  // TODO: use the given type hint for typeless textures
+  SCOPED_TIMER("VkDebugManager::PixelHistory");
+
+  if(sampleIdx > (uint32_t)imginfo.samples)
+    sampleIdx = 0;
+
+  uint32_t sampleMask = ~0U;
+  if(sampleIdx < 32)
+    sampleMask = 1U << sampleIdx;
+
+  bool multisampled = (imginfo.samples > 1);
+
+  if(sampleIdx == ~0U || !multisampled)
+    sampleIdx = 0;
+
+  // Replay up to (and not including) the first event.
+  m_pDriver->ReplayLog(0, events[0].eventId, eReplay_WithoutDraw);
+  m_pDriver->SubmitCmds();
+  m_pDriver->FlushQ();
+
+  RDCASSERT(m_pDriver->GetDeviceFeatures().occlusionQueryPrecise);
+  VkPhysicalDeviceProperties props = m_pDriver->GetDeviceProps();
+  VkPhysicalDeviceFeatures features = m_pDriver->GetDeviceFeatures();
+
+  VkQueryPool occlusionPool;
+  CreateOcclusionPool(dev, (uint32_t)events.size(), &occlusionPool);
+
+  PixelHistoryResources resources = {};
+  GetDebugManager()->PixelHistorySetupResources(resources, imginfo.extent, imginfo.format,
+                                    (uint32_t)events.size());
+
+  VulkanPixelHistoryCallback cb(
+      m_pDriver, x, y, (VkImage)GetResourceManager()->GetCurrentResource(target), imginfo.format,
+      sampleMask, occlusionPool, resources.colorImageView, resources.stencilImageView,
+      resources.colorImage, resources.stencilImage, resources.dstBuffer, events);
+  m_pDriver->ReplayLog(events.front().eventId, events.back().eventId, eReplay_Full);
```

Comment:
Since the callback already ignores other events than the ones it cares about, I'd suggest making this replay from event 0, and removing the above 'replay to first event'.

Generally speaking replaying from event X to event Y is not supported on vulkan especially when it can cross command buffer or submission boundaries. There's limited support when it's within the same command buffer but that won't always be the case here.

Reply:
Done.